src/compiler/checker.ts

```diff
@@ -30498,6 +30503,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             return type;
         }
 
+        const originalType = type;
         type = getNarrowableTypeForReference(type, node, checkMode);

```

Comment:
there are 2 other calls to `getNarrowableTypeForReference` that are now not accompanied by this `recombineUnknownType` dance. Before landing this, I should recheck both to see if that's OK. 

One of them is in `checkReturnExpression` - @gabritto do you have an opinion about this? This fix improves this scenario:
```ts
// this one doesn't work without this fix
function test1<T extends unknown>(
  x: T,
): T extends {} ? {} : T extends Nullable ? Nullable : never {
  if (x == undefined) {
    return x;
  } else {
    return x;
  }
}

type Nullable = null | undefined;

// this one works
function test2<T extends {} | Nullable>(
  x: T,
): T extends {} ? {} : T extends Nullable ? Nullable : never {
  if (x == undefined) {
    return x;
  } else {
    return x;
  }
}
```

but since the returned type can't "escape" the return statement, it feels redundant to recombine that unknown type there